### DISCUSSION TITLE
Async Communication

### DISCUSSION THREAD

**Discussion by paulheinr** [rsmpi/rsmpi#190] [Q&A] [answered]
Hi everyone,

I am trying to implement asynchronous messaging by implementing a simple client-server messaging. So there is one process sending requests to another, which is busy waiting and responding. I am using `immediate_send` for this.

I want to use async communication to flexibly send async requests, continue with the computations, possibly send other requests, and as soon as I get responses, process them. 

(1) As far as I understand the usage of `mpi::request::scope` (or `multiple_scope`) and `immediate_send`, the input variables (scope and buffer) and the return value (`Request`) must fullfill the same life time bounds. Since the scope object is constructed outside the callback, there is no other way than to construct the buffer outside the scope (as stated in the code documentation).

(2) Also, each `Request` must be completed before it can be dropped safely. So before we jump out of the callback, we have to wait for it.

In my opinion, both requirements make it very hard to do computations right after the send without waiting for it to complete. Because of condition 2, the scope needs to be created as early as possible, but because of condition 1, the scope needs to be created as late as possible. 
I couldn't think of a nice architecture to do the computations in between. In all of the examples, which are provided in this repository, it is waited for the send requests to be completed, before something else is computed.

Do you have an idea how to tackle this problem?

**Comment by jedbrown** [comment, marked as answer]
See `examples/immediate_multiple_test.rs` for a relevant example. 

Note that the `Request<'a>` uses the buffer's borrow lifetime `buf: &'a Buf`. This is necessary for correctness, otherwise the buffer could be reused before the communication completes (a race condition).

I can't see your test code, but I think you could pass the `Scope` in and return the `Request` if you don't want to wait for it yet.

**Reply by paulheinr** [reply]
Thank you very much for your reply! 

I think I found a way to solve my problem with the drawback of creating memory leaks. It should be possible to deallocate the memory manually, but that is beyond my question.

Here is the code:
```rust
use mpi::point_to_point::{Destination, Source};
use mpi::Rank;
use mpi::request::{Request, Scope};
use mpi::topology::{Communicator, SimpleCommunicator};

fn send<'a, S: Scope<'a> + Copy>(
    world: &SimpleCommunicator,
    scope: S,
    x: &'a i32,
) -> Request<'a, i32, S> {
    world.process_at_rank(1 as Rank).immediate_send(scope, x)
}

fn main() {
    let iter = 10;
    let universe = mpi::initialize().unwrap();
    let world = universe.world();

    if world.rank() == 0 {
        mpi::request::multiple_scope(iter as usize, |scope, req| {
            for i in 0..iter {
                let b = Box::new(i);
                let static_ref: &'static i32 = Box::leak(b);

                let r = send(&world, scope, static_ref);
                req.add(r);
                println!("sent message {}", i);

                //do some computations, which affect the next message
            }
            req.wait_all(&mut vec![]);
        });
    } else {
        let mut received = Vec::with_capacity(iter as usize);
        for _ in 0..iter {
            let mut buf = 0;
            world.process_at_rank(0 as Rank).receive_into(&mut buf);
            println!("received message {}", buf);
            received.push(buf);
        }
        println!("received all messages");
        for i in 0..iter {
            assert_eq!(received.get(i as usize).unwrap(), &i);
        }
    }
}
```

**Reply by jedbrown** [reply]
You can move the buffer storage outside the scope, as in
```rs
        let bufs: Vec<i32> = (0..iter).collect();
        mpi::request::multiple_scope(iter as usize, |scope, req| {
            for i in 0..iter {
                let r = send(&world, scope, &bufs[i as usize]);
```

**Reply by paulheinr** [reply]
Yes. But in my particular use case, I don't know all the messages in advance. It is not possible to dynamically push elements into the vector during the for loop and reference them for sending.

**Reply by jedbrown** [reply]
Here's another way:
```rs
        let bufs = vec![std::cell::OnceCell::new(); iter];
        mpi::request::multiple_scope(iter as usize, |scope, req| {
            for i in 0..iter {
                let b = bufs[i].get_or_init(|| i as i32);
                let r = send(&world, scope, b);
```

**Reply by paulheinr** [reply]
Very nice, thank you!
